SimPEG/electromagnetics/time_domain/sources.py

```diff
@@ -20,111 +20,424 @@
 ###############################################################################
 
 
-class BaseWaveform(properties.HasProperties):
```

Comment:
these updates will unfortunately conflict with #1059 ... they are reasonably lightweight, but we will have to figure out how we want to navigate that.